Add map and cache event journal

mmedenjak opened the pull request:
Add event journal for tracking mutation events on map and cache.
The events are written in a ringbuffer (event journal) and
can be read from MapProxyImpl, CacheProxy, ClientMapProxy and
ClientCacheProxy.
The reading is done by first subscribing to the stream and
then calling the read method.

Depends on:
https://github.com/hazelcast/hazelcast-client-protocol/pull/87
https://github.com/hazelcast/hazelcast-client-protocol/pull/88

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9256/

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/cache/impl/journal/RingbufferCacheEventJournalImpl.java:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    @Override
+    public void writeEvictEvent(ObjectNamespace namespace, int partitionId, Data key, Object value) {
```
It can be better to note that different entries can be evicted on primary and on backup as a result of the same call. So for example when `cache#put` returns, its possible to see that primary evicted entry-A but backup evicted entry-B. This leads to see different journals between replicas.

mmedenjak replied:
So if the partition owner is terminated and the backup is promoted you will get different entries?

ahmetmircik replied:
right

mmedenjak replied:
Added description on `MapEventJournal`, `CacheEventJournal` and all the member and client proxies. Please check

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java:
```diff
@@ -110,6 +111,30 @@
     }
 
+    @Override
+    public <B extends Data> B toData(Object obj, DataType type) {
```
i remember @jerrinot had some objections to add a this kind of methods but not sure if he keeps them still. We opened a related issue previously: https://github.com/hazelcast/hazelcast/issues/10249

mmedenjak replied:
I found it easier to have the same journal code for OS and EE - otherwise EE would have to extend the journal implementation class and override some `toData` method and the EE code would have to instantiate this extension class. This seemed awkward.

mmedenjak replied:
@jerrinot can you check if moving this method to OS is ok?

jerrinot replied:
I am not very excited about fiddling with various data types in a regular serialization service.

mmedenjak replied:
I don't have a strong reason to keep it this way, I just thought it would be simpler than having two EE event journal classes overriding some `toOnHeapData()` method and instantiating those classes in EE code. What do you think? Is there a better way of getting on-heap `Data` in OS?

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/ReadResultSetImpl.java:
```diff
@@ -88,46 +108,68 @@ public int readCount() {
+    @Override
+    public long getSequence(int index) {
+        if (seqs == null) {
+            throw new UnsupportedOperationException("Sequence IDs are not available when the cluster version is lower than 3.9");
```
is unexpected cluster version the only reason to see sequences as null?

mmedenjak replied:
3.8 members don't send sequences - that's the only scenario

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9379/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9445/

mmedenjak commented:
run-lab-run

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9454/

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/cache/impl/journal/CacheEventJournalReadResultSetImpl.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Override
+    public void addItem(long seq, Object item) {
+        final InternalEventJournalCacheEvent e = serializationService.toObject(item);
```
As discussed, the deserialisation of `InternalEventJournalCacheEvent` would happen if we allowed the ringbuffer to store the event in serialised format (or if there was an off-heap ringbuffer). But I've hardcoded the in-memory format for event journal ringbuffers to `OBJECT` - so that call is a no-op. I'll switch to a cast to make it more explicit.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9486/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9487/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9492/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9497/

mmedenjak commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

mmedenjak commented:
@cangencer addressed all comments. Also added a note in `EventJournalConfig` that it is intended to be used with jet :  https://github.com/hazelcast/hazelcast/pull/10685/files#diff-7dd13850a4da1bc206aacde20033ff8aR32

devOpsHazelcast commented:
Test PASSed.

Donnerbart commented on hazelcast/src/main/resources/META-INF/services/com.hazelcast.DataSerializerHook:
```diff
@@ -30,3 +30,4 @@ com.hazelcast.internal.usercodedeployment.impl.UserCodeDeploymentSerializerHook
 com.hazelcast.aggregation.impl.AggregatorDataSerializerHook
 com.hazelcast.projection.impl.ProjectionDataSerializerHook
 com.hazelcast.config.ConfigDataSerializerHook
+com.hazelcast.journal.EventJournalDataSerializerHook
```
Missing newline at EOF (violates some POSIX rules, which is a reason for an according CheckStyle rule, which of course doesn't check this file)

tkountis commented on hazelcast/src/main/java/com/hazelcast/cache/impl/CacheProxy.java:
```diff
@@ -360,4 +369,61 @@ public boolean removePartitionLostListener(String id) {
+     *                                       configured for this cache
+     * @since 3.9
+     */
+    public EventJournalInitialSubscriberState subscribeToEventJournal(int partitionId) {
```
Maybe async by default ?

mmedenjak replied:
Hm, good idea.

tkountis commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheEventJournalReadTask.java:
```diff
@@ -0,0 +1,120 @@
+
+    @Override
+    public String getMethodName() {
+        return null;
```
Missed ?

tkountis commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheEventJournalSubscribeTask.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public Permission getRequiredPermission() {
+        return new CachePermission(parameters.name, ActionConstants.ACTION_READ);
```
Different type of action, maybe subscribe ? I believe we have something similar in the listeners

mmedenjak replied:
Hm, ok. Currently this is a read-only task but in the future it will modify state. I'll try and put it as `SUBSCRIBE`

mmedenjak replied:
Hm, there's `ACTION_LISTEN`. WDYT?

tkountis commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheEventJournalSubscribeTask.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Override
+    public String getMethodName() {
+        return null;
```
Missed ?

mmedenjak commented on hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientBasicCacheJournalTest.java:
```diff
@@ -0,0 +1,84 @@
+import javax.cache.Cache;
+import javax.cache.CacheManager;
+
+public class ClientBasicCacheJournalTest extends BasicCacheJournalTest {
```
You're right. I think I'll definitely need to do some work on the tests.

mmedenjak commented on hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/ClientCacheProxy.java:
```diff
@@ -472,6 +504,64 @@ public boolean removePartitionLostListener(String id) {
+    /**
+     * Subscribe to the event journal for this cache and a specific partition ID.
+     * The method will register the subscription so that the event journal knows which
+     * items have been read by this subscriber. It will return the subscription ID and
```
Very likely, comes from the fact that I write down the behaviour all over the place and then forget to update it when it is updated.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9527/

devOpsHazelcast commented:
Test PASSed.
